### PR TITLE
Percentage alert, rolling window, threshold alert

### PR DESCRIPTION
**Pull request by JotaBlanco**
Hey @tomas-quix , let me know when you get on with this al I'll walk you through

### CODE REVIEW

File: python/model/Percentage-Alert/main.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# temporary (needed for dev)
+client.api_url = "https://portal-api.dev.quix.ai"
+
+# Change consumer group to a different constant if you want to run model locally.
+print("Opening input and output topics")
+
+# Define environmental variables
+input_topic = client.open_input_topic(os.environ["Quix__Workspace__Id"]+"-"+os.environ["input"], "default-consumer-group")
+output_topic = client.open_output_topic(os.environ["Quix__Workspace__Id"]+"-"+os.environ["output"])
+parameter_name = str(os.environ["ParameterName"])
```
Comment (tomas-quix) [inline python/model/Percentage-Alert/main.py:18]: Move ParameterName and percentage to inner class ctor please so we explain how we got it.

File: python/model/Percentage-Alert/main.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+perc_points_alert = float(os.environ["PercentagePointsAlert"])
+
+
+# Callback called for each incoming stream
+def read_stream(input_stream: StreamReader):
+
+    # Create a new stream to output data
+    output_stream = output_topic.create_stream(input_stream.stream_id + "-percentage-alert")
```
[the remark this answers is not in the document]
Comment (JotaBlanco) [reply python/model/Percentage-Alert/main.py]: Done

File: python/model/Percentage-Alert/main.py (hunk 1 above)
Comment (tomas-quix) [reply python/model/Percentage-Alert/main.py:18]: Do it for other projects as well, please.

File: python/model/Percentage-Alert/main.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+from quixstreaming.app import App
+from percentage_function import PercentageAlert
+import os
+
+
+# Create a client. The client helps you to create input reader or output writer for specified topic.
+client = QuixStreamingClient('eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IlpXeUJqWTgzcXotZW1pUlZDd1I4dyJ9.eyJodHRwczovL3F1aXguYWkvb3JnX2lkIjoicXVpeGRldiIsImh0dHBzOi8vcXVpeC5haS9yb2xlcyI6IiIsImlzcyI6Imh0dHBzOi8vYXV0aC5kZXYucXVpeC5haS8iLCJzdWIiOiJhdXRoMHw2YjljOGI1YS1kNGY0LTQzMDUtOTdlYS1hYzQ2ZmI2OGUzODUiLCJhdWQiOlsiaHR0cHM6Ly9wb3J0YWwtYXBpLmRldi5xdWl4LmFpLyIsImh0dHBzOi8vcXVpeC1kZXYuZXUuYXV0aDAuY29tL3VzZXJpbmZvIl0sImlhdCI6MTY0MTkwMTg5NCwiZXhwIjoxNjQ0NDkzODk0LCJhenAiOiI2MDRBOXE1Vm9jWW92b05Qb01panVlVVpjRUhJY2xNcyIsInNjb3BlIjoib3BlbmlkIHByb2ZpbGUgZW1haWwiLCJwZXJtaXNzaW9ucyI6W119.vpqLcoz023O3ZlddoD1Gcglm-j-btFmphrt_ur4wCMoejRo8tqbOTIDBkFRBlaL0AiduISoNBibevAu58Zx_riRoZf-KiZacehNSZHEBsk3qeiLPowEPCzJUFgiZW-LcHly-rfMzoLellrp7FKTcfu5Hg9hnkAHbWGKFeczjblElRuL4pLPcHiwNNwXlIHxm0KzegFgad7-JMKzdiKCwYruNINS83Ez0HGbsXbpDsri_0WGmXPYeX8yIbhimW0ZnQ3ZAaTwln22E87YbhE3bj4zzjHidBqIOXvdkbe8DRquPjdaEL8kvTp0Fm2Np2JKSagOteuQ4z5lLBMaj91g8Gw')
```
Comment (tomas-quix) [inline python/model/Percentage-Alert/main.py:8]: You have left token here.

File: python/model/Percentage-Alert/main.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    # handle the data in a function to simplify the example
+    quix_function = PercentageAlert(input_stream, output_stream, parameter_name, perc_points_alert)
+        
+    # React to new data received from input topic.
+    input_stream.events.on_read += quix_function.on_event_data_handler
+    input_stream.parameters.on_read += quix_function.on_parameter_data_handler
+
+    # When input stream closes, we close output stream as well. 
+    def on_stream_close(endType: StreamEndType):
```
Comment (tomas-quix) [inline python/model/Percentage-Alert/main.py:37]: endType should be end_type

File: python/model/Percentage-Alert/percentage_function.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
+from quixstreaming import StreamReader, StreamWriter, EventData, ParameterData
+import pandas as pd
+
+# Initiate global variables
+global_max = None
```
Comment (tomas-quix) [inline python/model/Percentage-Alert/percentage_function.py:5]: This is wrong, you need to use class fields. Will explain on a call.

File: python/model/Percentage-Alert/percentage_function.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                global_max_ti = ti
+
+
+    # Callback triggered for each new event.
+    def on_event_data_handler(self, data: EventData):
+        print(data)
+
+     # Callback triggered for each new parameter data.
+    def on_parameter_data_handler(self, data: ParameterData):
+        global global_max, global_min, global_max_ti, global_min_ti
```
Comment (tomas-quix) [inline python/model/Percentage-Alert/percentage_function.py:51]: This is wrong, you need to use class fields. Will explain on a call.

File: python/model/Threshold-Alert/main.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+from quixstreaming import QuixStreamingClient, StreamEndType, StreamReader
+from quixstreaming.app import App
+from threshold_function import threshold
+import os
+
+# Create a client. The client helps you to create input reader or output writer for specified topic.
+client = QuixStreamingClient(
+    'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IlpXeUJqWTgzcXotZW1pUlZDd1I4dyJ9.eyJodHRwczovL3F1aXguYWkvb3JnX2lkIjoicXVpeGRldiIsImh0dHBzOi8vcXVpeC5haS9yb2xlcyI6IiIsImlzcyI6Imh0dHBzOi8vYXV0aC5kZXYucXVpeC5haS8iLCJzdWIiOiJhdXRoMHw2YjljOGI1YS1kNGY0LTQzMDUtOTdlYS1hYzQ2ZmI2OGUzODUiLCJhdWQiOlsiaHR0cHM6Ly9wb3J0YWwtYXBpLmRldi5xdWl4LmFpLyIsImh0dHBzOi8vcXVpeC1kZXYuZXUuYXV0aDAuY29tL3VzZXJpbmZvIl0sImlhdCI6MTY0MTgwODE4MywiZXhwIjoxNjQ0NDAwMTgzLCJhenAiOiI2MDRBOXE1Vm9jWW92b05Qb01panVlVVpjRUhJY2xNcyIsInNjb3BlIjoib3BlbmlkIHByb2ZpbGUgZW1haWwiLCJwZXJtaXNzaW9ucyI6W119.ISUihdFG2rVCBXG_wjEbJ2fTYXwECX_QiN29IFITWTLvfjUSDHZeNCHdyN78oL4leGxTAUbVN7NPVlP_edV68Z-VGR6AhvDpXHZ8YQZkoSUMngcac0wTW2VHBIsBAyTCxD2ucQrt73m_P1rP_apQYkhpaPzA5RIK90anNvRLh59kIYlDYkrox6gyP_OA3pntnCTDdvmtGfHdt3z1dWz4EkbqjlMjVVNadxd2t1-mtQTPV1VbMZd-0CIUDb56HasOV9jaEBwpP894pKvHej6qGGrzuVBc7Kb4s2-2DXMNulifEUQOIJsgvJLkFBn_PsyDLFZmHRPA5yIDbulK-ZoivQ')
```
Comment (tomas-quix) [inline python/model/Threshold-Alert/main.py:8]: You have left the token here.

File: python/model/Threshold-Alert/main.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+# temporary (needed for dev)
+client.api_url = "https://portal-api.dev.quix.ai"
```
Comment (tomas-quix) [inline python/model/Threshold-Alert/main.py:10]: remove this.

**Comment by JotaBlanco** [comment]
Hi @tomas-quix, check it now

File: python/model/Threshold-Alert/main.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+output_topic = client.open_output_topic(os.environ["Quix__Workspace__Id"]+"-"+os.environ["output"])
+
+
+# Callback called for each incoming stream
+def read_stream(input_stream: StreamReader):
+    # Create a new stream to output data
+    output_stream = output_topic.create_stream(input_stream.stream_id + '-' + os.environ["Quix__Deployment__Name"])
+    output_stream.properties.parents.append(input_stream.stream_id)
+
+    print(type(output_stream))
```
Comment (tomas-quix) [inline python/model/Threshold-Alert/main.py:23]: Why?

**Pull request merged**